### ISSUE TITLE
[adherents] regression in v13 : date_format() returns a string

### ISSUE THREAD

**Issue by daamien** [Dolibarr/dolibarr#17986]
A regression was introduced to the adherent module stats in version 13 because of commit 7dd69a2b2b0cb2aa6009491e57cb87022b5156f0

The bug appears multiple times with the `date_format(...)` comparisons inside:

htdocs/adherents/class/adherentstats.class.php

see the diff:
https://github.com/Dolibarr/dolibarr/commit/7dd69a2b2b0cb2aa6009491e57cb87022b5156f0#diff-766aa34789b0256eca402f4f8f53c70b0f428bd830191baeff2935a6b5614054L95

The `date_format()` SQL function returns a string:

https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_date-format

Thus the following syntax is wrong 

```php
$sql .= " WHERE date_format(p.dateadh,'%Y') = ".((int) $year);
````

because we're comparing a string to an integer value. MySQL has permissive type conversion and it will not produce an error on this. However PostgreSQL is more strict and does not allow theses comparisons. 

If you want to extract an integer value from a date, it is better to use the `EXTRACT()` function 

```sql
SELECT EXTRACT(YEAR FROM now()) = 2021;
````

or even simpler in this case, there's a `YEAR()` function

```sql
SELECT YEAR(now()) = 2021;
````

I would recommend using MySQL strict mode in the CI tests to detect these kind of regressions early on.

https://dev.mysql.com/doc/refman/8.0/en/sql-mode.html

**Comment by daraelmin** [comment]
May be this would do the trick.

```php
$sql .= " WHERE date_format(p.dateadh,'%Y') = '".((int) $year)."'";
````

It force the string type in sql mode, but it avoid sql injection in $year.

**Comment by daraelmin** [comment]
However, I don't know witch solution is the best for perf.

**Comment by daamien** [comment]
The problem here is the use of the `date_format()` function.

We want to do a numeric comparison, so we need to extract a numeric value from `dateadh`

```php
$sql .= " WHERE YEAR(p.dateadh) = ". (int) $year;
```

**Comment by daamien** [comment]
See my comment in #17987

**Comment by daamien** [comment]
fixed by #17987
